src/concurrent.rs

```diff
@@ -0,0 +1,169 @@
+/// moved across threads.
+pub struct ShareableFile {
+    /// Underlying file descriptor shared across all views of this file.
+    fd: unix_io::RawFd,
+
+    /// Whether this instance of the file is responsible for closing the file descriptor.
+    owned: bool,
+
+    /// Whether the file descriptor has already been closed or not.
+    ///
+    /// In principle, we don't need this field: concurrent readers of this file will get their
+    /// reads abruptly terminated with an error, and we should be able to rely on EBADFD to tell
+    /// that this happened because of us closing the file descriptor.  Unfortunately... macOS
+    /// Mojave doesn't cooperate and yields very strange error codes (like EISDIR) pretty frequently
+    /// (but not deterministically).  May it be an APFS bug?
+    closed: Arc<AtomicBool>,
+}
+
+impl ShareableFile {
+    /// Constructs a new `ShareableFile` from an open file and takes ownership of it.
+    pub fn from(file: fs::File) -> ShareableFile {
+        use std::os::unix::io::IntoRawFd;
+        ShareableFile {
+            fd: file.into_raw_fd(),
+            owned: true,
+            closed: Arc::from(AtomicBool::new(false)),
+        }
+    }
+
+    /// Returns an unowned view of the file.
+    ///
+    /// Users of this file must accept that the file can be closed at any time by the owner.
+    pub fn clone_unowned(&mut self) -> ShareableFile {
+        ShareableFile {
+            fd: self.fd,
+            owned: false,
+            closed: self.closed.clone(),
+        }
+    }
+}
+
+impl Drop for ShareableFile {
+    fn drop(&mut self) -> () {
+        if self.owned {
+            debug!("Closing ShareableFile with fd {}", self.fd);
+
+            // We must mark the file as closed before attempting to close it because, when we do
+            // the actual close, any threads blocked on a read will get unlocked immediately and
+            // we won't have a chance to update the status.  It doesn't matter if the close succeeds
+            // or not though, because we cannot do anything useful in the latter case anyway.
+            self.closed.store(true, Ordering::SeqCst);
+
+            match unistd::close(self.fd) {
+                Ok(()) => (),
+                Err(e) => warn!("Failed to close fd {}: {}", self.fd, e)
+            }
+        }
+    }
+}
+
+impl io::Read for ShareableFile {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match unistd::read(self.fd, buf) {
+            Ok(n) => Ok(n),
+            Err(nix::Error::Sys(errno)) => {
+                if self.closed.load(Ordering::SeqCst) {
+                    Ok(0)  // Simulate EOF due to close in another thread.
+                } else {
+                    Err(io::Error::from_raw_os_error(errno as i32))
+                }
+            },
+            err => panic!("Did not expect to get an error without an errno from a nix call: {:?}",
```

Comment:
```suggestion
            Err(err) => panic!("Did not expect to get an error without an errno from a nix call: {:?}",
```

I think this makes it a bit more obvious which cases you are handling.

Reply:
I agree, but I'm wondering why I got a clippy warning yesterday with something similar... meh, I don't get it now, so the fixed version it is.